cmd/honeycloudtrail/main.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudtrail"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/honeycombio/honeyaws/logbucket"
+	"github.com/honeycombio/honeyaws/options"
+	"github.com/honeycombio/honeyaws/publisher"
+	"github.com/honeycombio/honeyaws/state"
+	libhoney "github.com/honeycombio/libhoney-go"
+	flag "github.com/jessevdk/go-flags"
+)
+
+var (
+	opt        = &options.Options{}
+	BuildID    string
+	versionStr string
+)
+
+// checks to see if struct contains a field
+// Trail type will not contain certain fields we want to check
+// such as S3KeyPrefix and S3BucketName if they are not set
+// causing a panic
+func nilField(v interface{}, name string) bool {
+	value := reflect.ValueOf(v)
+	if value.Kind() == reflect.Ptr {
+		value = value.Elem()
+	}
+	if value.Kind() != reflect.Struct {
+		return false
+	}
+
+	return value.FieldByName(name).IsNil()
+}
+
+func init() {
+	// set the version string to our desired format
+	if BuildID == "" {
+		versionStr = "dev"
+	} else {
+		versionStr = BuildID
+	}
+
+	// init libhoney user agent properly
+	libhoney.UserAgentAddition = "honeycloudtrail/" + versionStr
+}
+
+func cmdCloudTrail(args []string) error {
+	// TODO: Would be nice to have this more highly configurable.
+	//
+	// Will just use environment config right now, e.g., default profile.
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+
+	cloudtrailSvc := cloudtrail.New(sess, nil)
+
+	listTrailsResp, err := cloudtrailSvc.DescribeTrails(&cloudtrail.DescribeTrailsInput{})
+
+	if err != nil {
+		return err
+		os.Exit(1)
+	}
+
+	if len(args) > 0 {
+		switch args[0] {
+		case "ls", "list":
+			for _, trailSummary := range listTrailsResp.TrailList {
+				fmt.Println(*trailSummary.Name)
+			}
+			return nil
+
+		case "ingest":
+			if opt.WriteKey == "" {
+				logrus.Fatal(`--writekey must be set to the proper write key for the Honeycomb team.
+Your write key is available at https://ui.honeycomb.io/account`)
+			}
+
+			trailNames := args[1:]
+
+			if len(trailNames) == 0 {
+				for _, trail := range listTrailsResp.TrailList {
+					trailNames = append(trailNames, *trail.Name)
+				}
+			}
+
+			trailListResp, err := cloudtrailSvc.DescribeTrails(&cloudtrail.DescribeTrailsInput{
+				TrailNameList: aws.StringSlice(trailNames),
+			})
+			if err != nil {
+				fmt.Fprintln(os.Stderr, "Error getting trail descriptions: ", err)
+				os.Exit(1)
+			}
+
+			if len(trailListResp.TrailList) == 0 {
+				logrus.Fatal(`No valid trails listed. Try using ls to list available trails or refer to the README.`)
+				os.Exit(1)
+			}
+
+			var stater state.Stater
+
+			if opt.BackfillHr < 1 || opt.BackfillHr > 168 {
+				logrus.WithField("hours", opt.BackfillHr).Fatal("--backfill requires an hour input between 1 and 168")
+			}
+
+			if opt.HighAvail {
+				svc := dynamodb.New(sess)
```

Comment:
Tangential, but maybe this dynamodb setup code could be moved to `state.NewDynamoDBStater` (which would be extended to return an `err` as well) so it could be shared across the `cmd/*` binaries.